src/VisualStudio/Core/Def/Implementation/IntellisenseControls/IntellisenseTextBoxViewModelFactory.cs

```diff
@@ -0,0 +1,129 @@
+using Microsoft.VisualStudio.Text.Projection;
+using Microsoft.VisualStudio.TextManager.Interop;
+using Microsoft.VisualStudio.Utilities;
+using ImportingConstructorAttribute = System.Composition.ImportingConstructorAttribute;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.IntellisenseControls
+{
+    [Export(typeof(IntellisenseTextBoxViewModelFactory)), Shared]
+    internal class IntellisenseTextBoxViewModelFactory
+    {
+        private static readonly INITVIEW[] s_InitViews = new[] {
+            new INITVIEW()
+            {
+                fSelectionMargin = 0,
+                fWidgetMargin = 0,
+                fDragDropMove = 0,
+                IndentStyle = vsIndentStyle.vsIndentStyleNone
+            }
+        };
+
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly ITextEditorFactoryService _textEditorFactoryService;
+        private readonly OLE.Interop.IServiceProvider _serviceProvider;
+        private readonly IProjectionBufferFactoryService _projectionBufferFactoryService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public IntellisenseTextBoxViewModelFactory(
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            ITextEditorFactoryService textEditorFactoryService,
+            IContentTypeRegistryService contentTypeRegistryService,
+            IProjectionBufferFactoryService projectionBufferFactoryService,
+            SVsServiceProvider services)
+        {
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+            _textEditorFactoryService = textEditorFactoryService;
+            _serviceProvider = (OLE.Interop.IServiceProvider)services.GetService(typeof(OLE.Interop.IServiceProvider));
+            _projectionBufferFactoryService = projectionBufferFactoryService;
+        }
+
+        public async Task<IntellisenseTextBoxViewModel[]> CreateIntellisenseTextBoxViewModelsAsync(
+          Document document,
+          IContentType contentType,
+          string contentString,
+          Func<IProjectionSnapshot, ITrackingSpan[]> createSpansMethod,
+          string[][] rolesCollections, CancellationToken cancellationToken)
+        {
+            var vsTextLines = _editorAdaptersFactoryService.CreateVsTextBufferAdapter(_serviceProvider, contentType) as IVsTextLines;
+            vsTextLines.InitializeContent(contentString, contentString.Length);
+
+            var originalContextBuffer = _editorAdaptersFactoryService.GetDataBuffer(vsTextLines);
+            // Get the workspace, and from there, the solution and document containing this buffer.
+            // If there's an ExternalSource, we won't get a document. Give up in that case.
+            var solution = document.Project.Solution;
+
+            // Wrap the original ContextBuffer in a projection buffer that we can make read-only
```

Comment:
We've already cloned the buffer by calling InitializeContent on. Why do we need to create a projection buffer atop that rather than just redirectly marking that buffer as readonly?

Reply:
Sorry, I do not have an exact explanation. The code was borrowed. I just tried to use originalContextBuffer instead of contextBuffer (and remove making it readonly). It does not work.